sdks/common/runner-api/src/main/proto/beam_runner_api.proto

```diff
@@ -0,0 +1,450 @@
+  //
+  // Specifically, the URN should be enough to decipher the inputs,
+  // outputs, and payload.
+  string urn = 2;
+
+  // (Optional) if this node is contained within a composite, a pointer to the
+  // parent.
+  int64 parent_ref = 3;
+
+  // (Required) A map of local input names to target definitions.
+  //
+  // The URN for the graph node may clarify the type of the inputs
+  // (resp. outputs). For example:
+  //
+  //  - in the Runner API these are PCollections
+  //  - in the Fn API they may be Grpc ports
+  //
+  // The payload for this graph node may clarify the relationship of these
+  // inputs. For example:
+  //
+  //  - for a Flatten transform they are merged
+  //  - for a ParDo transform, some may be side inputs
+  //
+  // All inputs are recorded here so that the topological ordering of
+  // the graph is consistent whether or not the payload is understood.
+  //
+  // TODO: fix the value to be int64 as they will always be refs?
+  map<string, google.protobuf.Any> inputs = 4;
+
+  // (Required) A map from local output name to PCollection definitions.
+  //
+  // The URN or payload for the graph node may clarify the type and
+  // relationship of these. For example:
+  //
+  //  - for a ParDo transform, these are tags on PCollections, which will be
+  //    embedded in the DoFn.
+  //  - in the Runner API the targets are PCollections
+  //  - in the Fn API the targets may be Grpc ports
+  //
+  // TODO: fix the value to be int64 as they will always be refs?
+  map<string, google.protobuf.Any> outputs = 5;
+
+  // (Required) A payload determined by the URN, fully specifying this node.
+  google.protobuf.Any payload = 6;
+}
+
+// A PCollection!
+message PCollection {
+
+  // (Required) A pipeline-scoped unique id for this node.
+  int64 id = 1;
+
+  // (Required) A unique program-readable name for the value, such as
+  // "foo/baz/bar.out"
+  string label = 2;
+
+  // (Required) The id of the Coder for this PCollection.
+  int64 coder_ref = 3;
+
+  // (Required) Whether this PCollection is bounded or not
+  bool is_bounded = 4;
+
+  // (Required) The id of the windowing strategy for this PCollection.
+  int64 windowing_strategy_ref = 5;
+
+  // (Optional) Static display data for this PCollection.
+  DisplayData display_data = 6;
+}
+
+// An applied PTransform! This does not contain the graph data, but only the
+// fields specific to a graph node that is a Runner API transform
+// between PCollections.
+message PTransform {
+
+  // (Required) A pipeline-scoped unique id for this node.
+  int64 id = 1;
+
+  // (Required) A unique name for the application node.
+  //
+  // This should be stable over multiple runs of a driver program for the
+  // purposes of logging and associating pipeline state with a node.
+  string label = 2;
+
+  // (Optional) A URN that identifies what type of transform this is.
+  //
+  // If absent, this must be an "anonymous" composite transform.
```

Comment:
Is there any way to tell whether two "anonymous" composites are distinct from one-another? This could be useful, i.e. for display data. Perhaps each distinct anonymous composite should write a unique `urn` value.